File: src/Core.CSharp.Observe/EventLog.cs
using System.Numerics;

namespace Zeta.Core.CSharp.Observe;

/// <summary>
/// The observe event log as a <em>monoid</em>, made machine-recognized via
/// <c>System.Numerics</c> generic-math additive interfaces (B-0867.28 — the
/// interface-gate before building on top of the algebra). It wraps the
/// append-only list of <see cref="NextAction"/> events.
///
/// <para><b>Why additive-monoid (not <c>INumber&lt;T&gt;</c>)</b>: the event log is
/// the <em>free monoid</em> — identity = the empty log, operation = associative
/// append — and nothing more. It is not a field-like number (no multiplication,
/// no ordering, no negatives), so it implements only
/// <see cref="IAdditiveIdentity{TSelf, TResult}"/> +
/// <see cref="IAdditionOperators{TSelf, TOther, TResult}"/>, the exact algebraic
/// structure it has. (Decision: additive-monoidal only — operator 2026-05-31.)</para>
///
/// <para><b>The load-bearing law</b> is the homomorphism
/// <see cref="FoldOnto"/>: folding a concatenated log equals folding the second
/// onto the result of folding the first —
/// <c>(a + b).FoldOnto(w0) == b.FoldOnto(a.FoldOnto(w0))</c>. That is exactly the
/// append-only-replay-soundness / DST-replay property: re-folding a joined log
/// reproduces the same state as incremental folding. <c>Fold</c> is the monoid
/// action of (<see cref="EventLog"/>, +, empty) on <see cref="World"/>.</para>
///
/// <para>NOTE (consistent with <see cref="World"/>): the record's generated
/// equality compares <see cref="Events"/> by reference, not element-wise — value
/// comparison of two logs must compare <see cref="Events"/> with
/// <c>SequenceEqual</c>. UoM (<c>[&lt;Measure&gt;]</c>) is intentionally absent:
/// the observe <see cref="World"/> carries no measurable numeric quantity to
/// protect (only ids, flags, and a mode), so units here would be ceremony, not
/// safety; UoM lives where measurable quantities do (TriFloat values, the
/// attention/tick/dora domain).</para>
/// </summary>
/// <param name="Events">The append-only event log (MSB-of-history first).</param>
public sealed record EventLog(IReadOnlyList<NextAction> Events)
    : IAdditiveIdentity<EventLog, EventLog>,
      IAdditionOperators<EventLog, EventLog, EventLog>
{
    /// <summary>The empty log — the monoid identity. <c>empty + x == x == x + empty</c>.</summary>
    public static EventLog AdditiveIdentity { get; } = new(Array.Empty<NextAction>());

    /// <summary>Associative append — the free-monoid operation.</summary>
    public static EventLog operator +(EventLog left, EventLog right) =>
        new(left.Events.Concat(right.Events).ToList());

    /// <summary>Project this log onto an initial world: the monoid action /
    /// homomorphism to <see cref="World"/>. Equivalent to
    /// <c>Algebra.Fold(initial, Events)</c>; satisfies
    /// <c>(a + b).FoldOnto(w0) == b.FoldOnto(a.FoldOnto(w0))</c>.</summary>
    public World FoldOnto(World initial) => Algebra.Fold(initial, Events);
}

File: src/Core.FSharp.Observe/EventLog.fs
namespace Zeta.Core.FSharp.Observe

/// The observe event log as an additive MONOID (081KSXN940008QG0R0002287MP, Target 2 — F# slice),
/// expressed via F#'s native generic-math convention: `Zero` (identity) + `(+)`
/// (associative op). This is the F# parallel to the C# `EventLog`'s
/// `System.Numerics` `IAdditiveIdentity` + `IAdditionOperators` — each language
/// owns the interface in its OWN idiom. F#'s generic numeric machinery (SRTP,
/// `List.sum`, `LanguagePrimitives.GenericZero`) recognizes `Zero`/`(+)`, so this
/// is the genuine F# generic-math surface, not a transplant of the C# IWSAM
/// interfaces (which would fight the language).
///
/// Additive-monoidal ONLY (operator 2026-05-31): the log is the FREE monoid —
/// identity = empty log, op = associative append — not a field-like number
/// (no multiplication, ordering, or negatives to invent).
///
/// Load-bearing law: `FoldOnto` is the monoid action / homomorphism onto `World`
/// — folding a concatenated log equals incremental folding:
///   `(a + b).FoldOnto w0  =  b.FoldOnto (a.FoldOnto w0)`
/// i.e. append-only / DST-replay soundness.
///
/// UoM (`[<Measure>]`) is intentionally absent: the observe `World` carries no
/// measurable numeric quantity (only ids, flags, a mode), so units here would be
/// ceremony, not safety; UoM lives at the TriFloat / attention value domain
/// (`src/Core/Units.fs`). F# records give structural equality (incl. the
/// `NextAction list`), so two logs with equal events are value-equal natively.
type EventLog =
    { Events: NextAction list }

    /// The empty log — the monoid identity (F# generic-math `Zero`).
    static member Zero: EventLog = { Events = [] }

    /// Associative append — the free-monoid operation.
    static member (+)(left: EventLog, right: EventLog) : EventLog =
        { Events = left.Events @ right.Events }

    /// Project this log onto an initial world: the monoid action / homomorphism
    /// to `World`. Equivalent to `Algebra.fold initial this.Events`; satisfies
    /// `(a + b).FoldOnto w0 = b.FoldOnto (a.FoldOnto w0)`.
    member this.FoldOnto(initial: World) : World = Algebra.fold initial this.Events
